packages/edit-site/src/components/page-templates/template-actions.js

```diff
@@ -0,0 +1,228 @@
+import isTemplateRevertable from '../../utils/is-template-revertable';
+import isTemplateRemovable from '../../utils/is-template-removable';
+import { TEMPLATE_POST_TYPE } from '../../utils/constants';
+
+export function useResetTemplateAction() {
+	const { revertTemplate } = useDispatch( editSiteStore );
+	const { saveEditedEntityRecord } = useDispatch( coreStore );
+	const { createSuccessNotice, createErrorNotice } =
+		useDispatch( noticesStore );
+	return useMemo(
+		() => ( {
+			id: 'reset-template',
+			label: __( 'Reset template' ),
+			isPrimary: true,
+			icon: backup,
+			isEligible: isTemplateRevertable,
+			async perform( template ) {
+				try {
+					await revertTemplate( template, { allowUndo: false } );
+					await saveEditedEntityRecord(
+						'postType',
+						template.type,
+						template.id
+					);
+
+					createSuccessNotice(
+						sprintf(
+							/* translators: The template/part's name. */
+							__( '"%s" reverted.' ),
+							decodeEntities( template.title.rendered )
+						),
+						{
+							type: 'snackbar',
+							id: 'edit-site-template-reverted',
+						}
+					);
+				} catch ( error ) {
+					const fallbackErrorMessage =
+						template.type === TEMPLATE_POST_TYPE
+							? __(
+									'An error occurred while reverting the template.'
+							  )
+							: __(
+									'An error occurred while reverting the template part.'
+							  );
+					const errorMessage =
+						error.message && error.code !== 'unknown_error'
+							? error.message
+							: fallbackErrorMessage;
+
+					createErrorNotice( errorMessage, { type: 'snackbar' } );
+				}
+			},
+		} ),
+		[
+			createErrorNotice,
+			createSuccessNotice,
+			revertTemplate,
+			saveEditedEntityRecord,
+		]
+	);
+}
+
+export const deleteTemplateAction = {
+	id: 'delete-template',
+	label: __( 'Delete template' ),
+	isPrimary: true,
+	icon: trash,
+	isEligible: isTemplateRemovable,
+	modalProps: {
+		ModalContent: ( { item: template, setIsModalOpen } ) => {
+			const { removeTemplate } = useDispatch( editSiteStore );
+			return (
+				<VStack spacing="5">
+					<Text>
+						{ sprintf(
+							// translators: %s: The template or template part's title.
+							__( 'Are you sure you want to delete "%s"?' ),
+							decodeEntities( template.title.rendered )
+						) }
+					</Text>
+					<HStack justify="right">
+						<Button
+							variant="tertiary"
+							onClick={ () => {
+								setIsModalOpen( false );
+							} }
+						>
+							{ __( 'Cancel' ) }
+						</Button>
+						<Button
+							variant="primary"
+							onClick={ () =>
+								removeTemplate( template, {
+									allowUndo: false,
+								} )
+							}
+						>
+							{ __( 'Delete' ) }
+						</Button>
+					</HStack>
+				</VStack>
+			);
+		},
+	},
+};
+
+export const renameTemplateAction = {
+	id: 'rename-template',
+	label: __( 'Rename' ),
+	isEligible: ( template ) =>
+		isTemplateRemovable( template ) && template.is_custom,
+	modalProps: {
```

Comment:
I thought we'd just use `perform` for this, no special API for modals. Related conversation https://github.com/WordPress/gutenberg/pull/55192#discussion_r1352966570 

The action itself renders the `Modal` component from `@wordpress/components` if they want. It'd be essentially the same code the existing [rename action](https://github.com/WordPress/gutenberg/blob/d9cb65db2f13c1aa0ce342fb3b2da282d45179fe/packages/edit-site/src/components/template-actions/rename-menu-item.js#L88).

Reply:
In my mind, at least for now, we should have an API that kind of restricts how the UI is, to keep it coherent. What this means is that all actions are rendered either with an icon(primary) or a menu item in the dropdown(secondary). But these are implementation details and could change.

So the rationale here is to start by only allowing to provide modal contents and keep control of how to handle the buttons UI.

Reply:
Yeah, I was hesitant too. The only way to allow rendering anything with the "perform" API is to actually render the modal outside the data views component which felt not great to me. 

The other thing which is not directly related is that maybe we should rename `perform` to `callback` to match the shape of the API we have for commands. (which are similar to actions in a lot of ways)